Untangle theme screens

scruffian opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

Fixes #

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
*

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Go to '..'
*

scruffian committed:
Add Theme Install page

scruffian committed:
changelog

scruffian committed:
add theme-install.php

scruffian committed:
add theme-install.php

ryelle committed:
Make theme-install load

scruffian committed:
add js file

draganescu committed:
don't send simple site users to calypso for installed or activated themes, send them to classic view (wp-admin)

draganescu committed:
changelog

ryelle committed:
Use the page load hook to enqueue script

ryelle committed:
Ensure the custom JS is loaded after the core theme JS

ryelle committed:
Update the router to use the custom page URL

draganescu committed:
remove the theme showcase banner and link

jeryj commented on projects/packages/jetpack-mu-wpcom/src/features/wpcom-themes/wpcom-themes.php:
```diff
@@ -7,70 +7,6 @@
-/**
- * Displays a banner before the theme browser that links to the WP.com Theme Showcase.
- */
-function wpcom_themes_show_banner() {
```
We can't remove this function until it's removed from being called on WPCOM: fbhepr%2Skers%2Sjcpbz%2Sjc%2Qpbagrag%2Snqzva%2Qcyhtvaf%2Sjcpbz%2Qgurzrf.cuc%3Se%3Q04155113%23176%2Q179-og

ryelle committed:
Copy over atomic theme api customization

ryelle committed:
Fork the core admin-ajax response to return regardless of install permissions

# Conflicts:
#	projects/packages/jetpack-mu-wpcom/src/features/wpcom-themes/wpcom-themes.php

ryelle committed:
Fix the URLs passed to the router

ryelle commented:
In https://github.com/Automattic/jetpack/pull/41374/commits/02fe4c22781c1e03a8b7bb0c6a5468c4d10b799e I copied over the files from wpcomsh (the code used for Atomic themes), but this doesn't remove those original files yet. Then in https://github.com/Automattic/jetpack/pull/41374/commits/32b49638c86c36deff198e26126ede51e863fdab, I forked the core `wp_ajax_query_themes` handler, because it [first checks for install_themes permissions](https://github.com/WordPress/wordpress-develop/blob/3fd93789b81fb6ca7fdb5fb2794f41177c967bf4/src/wp-admin/includes/ajax-actions.php#L3613-L3619). I don't like this approach, copying more core code, so I'm open to better solutions — but this at least lets the existing wpcom API functions hook into the correct results.

ryelle committed:
Replace the ajax request for the favorites username with patched version

ryelle committed:
Update comments to explicitly call out patched sections of ajax functions

ryelle committed:
Add filter to update the theme object before returning in ajax response

ryelle committed:
Use the correct theme slug for each environment

On simple sites, the `stylesheet` includes the path (e.g., `pub/`), while atomic sites don't use this folder structure, so the "installed" theme would not match. In this case, the `id` value is the correct slug.

ryelle committed:
Add patched notes and remove hardcoded changes

ryelle committed:
Add filters to the theme templates

ryelle committed:
Add (unstyled) theme tier label

ryelle commented:
There are now wrapper filters on the template code — `wpcom_themes_tmpl_theme` for the card template & `wpcom_themes_tmpl_theme_preview` for the full-screen customizer preview. I've then added in the theme tier, passed back from the wpcom endpoint, and added it next to the theme name. It's not styled, probably needs to go somewhere else, but it's starting point.

Screenshot:

<img width="1337" alt="Screenshot 2025-01-31 at 12 26 03 PM" src="https://github.com/user-attachments/assets/b4619e63-f571-48ee-bab0-52dece6ad2a4" />

ryelle committed:
Remove the tier check at the API level

matiasbenedetto committed:
Conditional theme activate button text depending of the user plan

ryelle committed:
Update "Activate" and "Install" URLs to use checkout if necessary

ryelle committed:
Update "Install" button text to "Upgrade to install", ensure only button text is replaced

ryelle commented:
I'm not sure the best place for this, but just to note this down somewhere, here's how I'm thinking about the work breakdown for the `theme-install.php` portion of the work (this does cross over a few of the other open issues). Unchecked checkboxes are tasks anyone can pick up, so we can work on this in parallel.

- Add a custom theme-install page for Simple sites
  - [x] New page on simple sites, should show themes & core filters
  - [ ] Add DIFM banner (IIRC this should support JiTM in general, which is in progress by @getdave)
  - [ ] Add “Upload theme” button to header, should reveal Atomic upsell
- Display wpcom, partner, and community themes
  - [ ] Community themes that are available on wpcom should be hidden, priority given to the wpcom theme (for example, searching "Layover" should only return the wpcom version) — probably not working due to the prefixed stylesheet value.
  - [ ] Themes should display plan label ([reference PR for text and conditions](https://href.li/?https://github.com/Automattic/wp-calypso/pull/98451)), buttons should change to “Upgrade to…” (in progress by @matiasbenedetto)
  - [ ] “Install" button tries to run behind the scenes with install ajax request, but fails (of course). This applies to community & partner themes.
  - [ ] Add partner themes to Atomic results.
- Users should be able to upgrade and activate paid themes/plans, including community and partner
  - [x] Personal & Premium are working great
  - [ ] Atomic conversion currently requires a visit to the Theme Showcase page ([see these notes](https://github.com/Automattic/wp-calypso/issues/98662#issuecomment-2608323961)). We talked about moving this screen to wp-admin; also simply redirecting back to wp-admin after theme activation.
  - [ ] Some partner themes (Drinkify) don't populate `product_details`, so they’re not purchasable.
- [ ] Direct links to theme should show wpcom theme info. For example, `/wp-admin/themes.php?page=wpcom-install-theme&theme=pub%2Fconference`. Not working on Atomic either (`/wp-admin/theme-install.php?theme=timestream`).

ryelle committed:
Fix duplicate ampersand in URL

ryelle committed:
Fix CSS on custom page by injected theme-install-php class

ryelle committed:
Switch installTheme handler to a noop, to prevent AJAX install request

ryelle committed:
Apply the install button fix on preview too

ryelle commented:
Eventually we'll need to figure out how to roll this out - not that it's ready yet, but to think about:

- All the `packages/jetpack-mu-wpcom/src/features/wpcom-themes/includes/*` files are simply copied from `plugins/wpcomsh/wpcom-themes/includes/`; but they should live in one unified place, for both Atomic and Simple
- The modified `theme-install.php` file is based on a core file. How will we handle that? Even if the filters we added are merged into core, Simple sites can't use it directly. So does it actually live forever as a forked file? Not that I see this as a file likely to change often, but still.

scruffian committed:
add an upload button

scruffian commented:
> So does it actually live forever as a forked file? 

I would think so yes. We'd need to copy/paste it every time we merge core in.

matiasbenedetto committed:
Display a text saying if the theme is available on the plan or the user needs to upgrade to another plan

matiasbenedetto committed:
copying badge styles from Gutenberg's Badge component

scruffian commented on projects/packages/jetpack-mu-wpcom/src/features/wpcom-themes/wpcom-themes.php:
```diff
@@ -7,85 +7,363 @@
+ * @param object $theme The theme object.
+ * @return object The theme object.
+ */
+function wpcom_themes_api_theme_object( $theme ) {
```
Should this get added to the endpoint itself?

ryelle replied:
I left it as a filter (in `wpcom_themes_ajax_query_themes`) because that function is copied from core, and our code here needs to run after the body of that function runs.

The alternative would be updating the endpoint, but then we would need to alter the `wpcom_themes_ajax_query_themes` more, which could be harder to merge later if there are core changes. I added a note about that to [my how to merge comment](https://github.com/Automattic/jetpack/pull/41374#issuecomment-2651696266).

scruffian committed:
fix a couple of PHP warnings

scruffian commented on projects/packages/jetpack-mu-wpcom/src/features/wpcom-themes/includes/class-wpcom-themes-api.php:
```diff
@@ -0,0 +1,137 @@
+	 *
+	 * @return array<stdClass> An array with all the WPCom themes.
+	 */
+	protected function fetch_themes( string $cache_key, array $params = array() ): array {
```
This function calls the API through HTTP. On Simple sites we can call the code directly, which means that we don't need to wait for an HTTP request. I think we can do this using Client::wpcom_json_api_request_as_blog

scruffian replied:
Done in [458df77](https://github.com/Automattic/jetpack/pull/41374/commits/458df77a84a7c2a51b44c19e5e2e999b0bbdad31) - we should now be able to modify the API and see the results when testing.

scruffian replied:
I reverted this because it was broken, but it should give you an idea how to do it!

scruffian commented:
> All the packages/jetpack-mu-wpcom/src/features/wpcom-themes/includes/* files are simply copied from plugins/wpcomsh/wpcom-themes/includes/; but they should live in one unified place, for both Atomic and Simple

I think the first step here could be to open a separate PR to move those files from wpcomsh to here. That will also make the diff easier to grok.

jeryj commented:
> I think the first step here could be to open a separate PR to move those files from wpcomsh to here. That will also make the diff easier to grok.

I can do this if no one else has yet.

ryelle committed:
Switch to the marketplace install URL

jeryj commented:
PR for moving files in the works: https://github.com/Automattic/jetpack/pull/41770/files

ryelle committed:
WPCom Theme merge: Merge by actual theme slug, not wpcom theme slug

Installed themes on wpcom have prefixed slugs like `pub/*` or `premium/*` due to the file structure, but wporg expects just the theme name (e.g. psychedeli).

ryelle committed:
WPCom Theme API: Prevent duplicate overrides on block themes request

ryelle commented:
The commits in https://github.com/Automattic/jetpack/pull/41374/commits/f188246a2ce7f991d6aee434ec61ad76987955f8 & https://github.com/Automattic/jetpack/pull/41374/commits/6d58cde8976f738830a1b7e28a830b06b93dd4b6 fix some API merge issues. The first ensures there are no duplicates in the results, so you can search for "Layover" or "Videomaker" and you won't see the community option. The second commit prevents `wpcomsh_feature_filter_wpcom_themes_api_result` from running over the "Block Themes" tab, since it's already running `wpcomsh_block_themes_wpcom_themes_api_result`. According to that function, it should merge the resulting themes by release date, but since both functions ran, the general filter result was merging with wpcom themes first. Now it's back to just using release date. Is this actually the preferred behavior?

There is an interesting … glitch? i guess — the `merge_by_release_date` function filters out any theme released after the first wporg theme, so right now on Latest, you'll see "Omega Food Delivery" as the most recent, even though technically "Booknest" was released more recently. I assume it's to handle paginated requests, but it should probably not apply to the first page of results.

These both change the `WPCom_Themes*` functions — should I move those over to #41770 or do we want to have a separate PR for changes to those files?

scruffian commented:
> These both change the WPCom_Themes* functions — should I move those over to https://github.com/Automattic/jetpack/pull/41770 or do we want to have a separate PR for changes to those files?

I think let's keep #41770 just for the file moving, so we can assess the impact of the changes we're making more carefully.
